### PR TITLE
[WIP] Register vpc id and subnet id attributes

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

Registers container instance with vpc and subnet ID attributes. The WIP part alludes to the changes needed to handle the case where the Agent is running on an instance in classic EC2, where we want to explicitly disable all the Task ENI bits, which has a dependency on the code in #779

### Implementation details

By extending the metadata interface to retrieve subnet and vpc id's

### Testing

Please note that functional tests fail with this PR as the `ecs.vpc-id` and `ecs.subnet-id` attributes are not whitelisted by the ECS backend. However, reverting the latest commit causes them to pause. I'll leave the latest commit it and will revert it before checking in. I'll re-revert that once those attributes are whitelisted.

- [X] Builds on Linux (`make release`)
- [X] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [X] Unit tests on Linux (`make test`) pass
- [X] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [X] Integration tests on Linux (`make run-integ-tests`) pass
- [X] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [X] Functional tests on Linux (`make run-functional-tests`) pass
- [X] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes:  yes

### Description for the changelog

None

### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

File: agent/api/ecsclient/client.go [hunk 1]
```diff
@@ -259,10 +259,37 @@ func getCpuAndMemory() (int64, int64) {
...
+	mac, err := client.ec2metadata.PrimaryENIMAC()
+	if err != nil {
+		seelog.Warnf("Unable to get the MAC address of the primary network interface: %v", err)
+		return attributes
+	}
+
+	vpcID, err := client.ec2metadata.VPCID(mac)
+	if err != nil {
+		seelog.Warnf("Unable to get the VPC ID of the primary network interface: %v", err)
+		return attributes
```
Comment (nmeyerhans) [inline agent/api/ecsclient/client.go]: What's the rationale behind immediately returning upon the first encountered error? It's conceivable (though unlikely) that retrieval of the VPCID fails but PrimaryENIMAC and SubnetID are successful. If this happens, our return value contains neither of the successfully retrieved attributes.

Comment (aaithal) [inline agent/api/ecsclient/client.go]: That's intentional. Though it's debatable if that's optimal. These attributes (at least for now) will be used for determining if an instance satisfies the networking parameters for the backend to create and attach the ENI. I wanted to ensure that either both subnet and vpc id get registered or neither of them do as that's a stronger protocol for the backend to deal with.

File: agent/api/ecsclient/client.go [hunk 2]
```diff
@@ -90,10 +87,10 @@ func (client *APIECSClient) SetSubmitStateChangeSDK(sdk api.ECSSubmitStateSDK) {
 func (client *APIECSClient) CreateCluster(clusterName string) (string, error) {
 	resp, err := client.standardClient.CreateCluster(&ecs.CreateClusterInput{ClusterName: &clusterName})
 	if err != nil {
-		log.Crit("Could not register", "err", err)
+		seelog.Criticalf("Could not create cluster, error: %v", err)
```
Comment (aaithal) [inline agent/api/ecsclient/client.go]: Will do.

File: agent/ec2/ec2_metadata_client.go [hunk 3]
```diff
@@ -15,25 +15,30 @@ package ec2
...
+	SecurityCrednetialsResource               = "/2014-02-25/meta-data/iam/security-credentials/"
+	InstanceIdentityDocumentResource          = "/2014-02-25/dynamic/instance-identity/document"
+	InstanceIdentityDocumentSignatureResource = "/2014-02-25/dynamic/instance-identity/signature"
+	SignedInstanceIdentityDocumentResource    = "/2014-02-25/dynamic/instance-identity/pkcs7"
+
+	macResource            = "/2014-02-25/meta-data/mac"
+	vpcIDResourceFormat    = "/2014-02-25/meta-data/network/interfaces/macs/%s/vpc-id"
+	subnetIDResourceFormat = "/2014-02-25/meta-data/network/interfaces/macs/%s/subnet-id"
+
+	EC2MetadataRequestTimeout = time.Duration(1 * time.Second)
```
Comment (richardpen) [inline agent/ec2/ec2_metadata_client.go]: What's the benefit of not using `1 * time.Second` directly?

File: agent/ec2/ec2_metadata_client.go [hunk 4]
```diff
@@ -61,100 +58,84 @@ type InstanceIdentityDocument struct {
...
 
-type ec2MetadataClientImpl struct {
-	client HttpClient
+type ec2MetadataClient struct {
+	httpClient http.Client
 }
 
-func NewEC2MetadataClient(httpClient HttpClient) EC2MetadataClient {
+// NewEC2MetadataClient creates a new ec2MetadataClient object
+func NewEC2MetadataClient(httpClient http.Client) EC2MetadataClient {
```
Comment (richardpen) [inline agent/ec2/ec2_metadata_client.go]: Is there any reason not using the `ec2metadata` in `aws-sdk-go`?

File: agent/ec2/ec2_metadata_client.go [hunk 5]
```diff
@@ -61,100 +58,84 @@ type InstanceIdentityDocument struct {
...
+			resp, err = c.httpClient.Get(endpoint)
+			if err == nil && resp.StatusCode == 200 {
+				return nil
+			}
+			if resp != nil && resp.Body != nil {
+				resp.Body.Close()
+			}
+			if err == nil {
+				seelog.Warnf("Error accessing the EC2 Metadata Service; non-200 response: %v", resp.StatusCode)
+				return fmt.Errorf("ec2 metadata client: unsuccessful response from Metadata service: %v", resp.StatusCode)
```
Comment (richardpen) [inline agent/ec2/ec2_metadata_client.go]: you can use `errors.Errorf` here.

File: agent/ec2/ec2_metadata_client.go [hunk 6]
```diff
@@ -61,100 +58,84 @@ type InstanceIdentityDocument struct {
...
-	if err != nil {
-		return nil, err
-	}
-	return &credential, nil
+	return &ec2MetadataClient{httpClient: httpClient}
 }
 
-func (c *ec2MetadataClientImpl) InstanceIdentityDocument() (*InstanceIdentityDocument, error) {
-	rawIidResp, err := c.ReadResource(INSTANCE_IDENTITY_DOCUMENT_RESOURCE)
+func (c *ec2MetadataClient) InstanceIdentityDocument() (*InstanceIdentityDocument, error) {
```
Comment (richardpen) [inline agent/ec2/ec2_metadata_client.go]: Let's get rid of one character variable, `client`?

File: agent/ec2/ec2_metadata_client.go (hunk 4 above)
Comment (aaithal) [inline agent/ec2/ec2_metadata_client.go]: I'm not sure why we had this in the first place. But, replacing this with the one in SDK is outside the scope of this PR.

File: agent/ec2/ec2_metadata_client.go (hunk 6 above)
Comment (aaithal) [inline agent/ec2/ec2_metadata_client.go]: That would be better. But, this is legacy code. I think the better thing to do is to get rid of most of this file itself. I think this is outside the scope of this PR as well.

File: agent/ec2/ec2_metadata_client.go (hunk 5 above)
Comment (aaithal) [inline agent/ec2/ec2_metadata_client.go]: sure

File: agent/ec2/ec2_metadata_client.go (hunk 3 above)
Comment (aaithal) [inline agent/ec2/ec2_metadata_client.go]: Legacy code. I can only guess. I think the reason would be to have all timeouts in one place so that we can easily change it if needed

Comment (aaithal) [inline agent/ec2/ec2_metadata_client.go]: Ah, I misunderstood, apologies. I'm not sure tbh. I'll modify this legacy code.

File: agent/ec2/ec2_metadata_client.go (hunk 4 above)
Comment (samuelkarp) [inline agent/ec2/ec2_metadata_client.go]: @aaithal We had this in the first place because this code is older than `ec2metadata` in `aws-sdk-go`. 
@richardpen Feel free to make an issue for getting rid of our custom one in favor of the one from the SDK.

**Comment by aaithal** [comment]
I'm closing this as some of the relevant code has significantly diverged. Will create a new PR by reusing some of these commits.
